gluon-core: split dual lan devices

AiyionPrime opened the pull request:
This allows to split port assignments of devices, that were previously distinct and are now part of one `switch_vlan`.

Following the discussion in the latest meetup,
a variant of this should either be merged to account for #2439 before the next release,
or the other way around (reverting #2439).

I'll rebase #2462 onto this to show my intent.

Currently the statement to check, whether recreating the switch_vlans is necessary is not implemented. If you've got suggestions how to do it properly?

AiyionPrime commented:
Currently running on http://acpro-lanwan.n.ffh.zone
http://[2001:678:978:214:b6fb:e4ff:fe40:3e83]/cgi-bin/status

blocktrron commented on package/gluon-core/luasrc/lib/gluon/upgrade/115-swconfig:
```diff
@@ -0,0 +1,42 @@
+local uci = require('simple-uci').cursor()
+
+local function get_keys(t)
+	local keys={}
+	local n=0
+
+	for k,_ in pairs(t) do
+		n=n+1
+		keys[n]=k
+	end
+	return keys
```
Consistent spacing for assignments.

AiyionPrime replied:
resolved in 7cab846e9ceed11755c29030864962bc345da2ef

blocktrron commented on package/gluon-core/luasrc/lib/gluon/upgrade/115-swconfig:
```diff
@@ -0,0 +1,42 @@
+
+	local assignments = switch_vlans[board_name]
+
+--	if device_required_vlan_config then
```
In case this should be a guard to only apply the workaround on first setup, this should happen at the beginning of the script.

OTOH, i see no downside in executing this script everytime

AiyionPrime replied:
In that case I'd keep it simple and go with the latter.

NeoRaider replied:
As the network config is reset on each reconfigure, there is no difference between the first and later runs.

blocktrron commented on package/gluon-core/luasrc/lib/gluon/upgrade/115-swconfig:
```diff
@@ -0,0 +1,42 @@
+local devices = get_keys(switch_vlans)
+
+
+if platform.match('ath79', 'generic', devices) then
```
How about `if not platform.match ... then return fi`

AiyionPrime replied:
resolved in 7cab846e9ceed11755c29030864962bc345da2ef

NeoRaider commented on package/gluon-core/luasrc/lib/gluon/upgrade/115-swconfig:
```diff
@@ -0,0 +1,40 @@
+	local n = 0
+
+	for k,_ in pairs(t) do
+		n = n+1
+		keys[n] = k
```
```suggestion
		table.insert(keys, k)
```
While using an explicit counter variable is slightly more efficient, this shouldn't matter here, and I consider `table.insert()` more idiomatic Lua.

AiyionPrime replied:
outdated

NeoRaider commented on package/gluon-core/luasrc/lib/gluon/upgrade/115-swconfig:
```diff
@@ -0,0 +1,40 @@
+
+uci:delete_all('network', 'switch_vlan')
+
+for vlan,ports in pairs(assignments) do
```
```suggestion
for vlan, ports in pairs(assignments) do
```

AiyionPrime replied:
resolved in 4fbc4e0cd17fb40269dc39f0525f9cab06c0e11c

NeoRaider commented on package/gluon-core/luasrc/lib/gluon/upgrade/115-swconfig:
```diff
@@ -0,0 +1,40 @@
+
+local devices = get_keys(switch_vlans)
+
+if not platform.match('ath79', 'generic', devices) then
```
Actually, we don't even need the `get_keys()` helper:
```suggestion
local board_name = platform.get_board_name()
local assignments = switch_vlans[board_name]
if not platform.match('ath79', 'generic') or not assignments then
```

AiyionPrime replied:
resolved in 4fbc4e0cd17fb40269dc39f0525f9cab06c0e11c

NeoRaider commented on package/gluon-core/luasrc/lib/gluon/upgrade/115-swconfig:
```diff
@@ -0,0 +1,40 @@
+uci:delete_all('network', 'switch_vlan')
+
+for vlan,ports in pairs(assignments) do
+	local wan_name = uci:add("network", "switch_vlan")
+	uci:set("network", wan_name, "device", "switch0")
+	uci:set("network", wan_name, "vlan", vlan)
+	uci:set("network", wan_name, "ports", ports)
```
```suggestion
	uci:section("network", "switch_vlan", nil, {
		device = "switch0",
		vlan = vlan,
		ports = ports,
	}
```

AiyionPrime replied:
resolved in 4fbc4e0cd17fb40269dc39f0525f9cab06c0e11c

AiyionPrime commented:
> Works fine, except the configmode is currently delivered on the wrong port (lan).

resolved in 36a10be3b77ae647f7e5ff8372208e20d57f83b9

compiling now testing done in 35 minutes

AiyionPrime reviewed (commented):
I'd be fine with this. Any takers?

AiyionPrime commented:
Updated commit message to start with `gluon-code:`.

NeoRaider commented:
> Updated commit message to start with `gluon-code:`.

Should be "gluon-core", not "gluon-code"

AiyionPrime commented:
Updated commit message to start with `gluon-core`.
I suppose that's fine with blocktrron as well.

Changed nothing else, and master did not change either.
